demodata/recipe.yaml

```diff
@@ -12,8 +12,12 @@ dimensions:
         CITY: Brno
     - environment:
         CITY: Boston
+      when:
+        - CONTEXT.color is match("blue")
   colors:
     - context:
         color: red
+      when:
+        - JOB.environment.CITY is not match("Brno")
```

Comment:
While for Jira issue config `when` is a string, here it seems more practical to have a list, since we are merging it. However, all conditions in such a list must match (AND), while users may intuitively understand it as any condition would be sufficient. WDYT? Should we stick to a single condition (string) and somehow rewrite the merging or should we rather support OR within a single `when` (before merging)?

Reply:
From Ansible we're used to treating the list of conditions as if an implicit `and` is applied to all items. `or` in a single condition should already be supported, together with `and` and brackets. I did not check for this, but even multiline conditions might work (needs to be tested), e.g.

```yaml
when: >-
  EVENT.type is erratum
  and (EVENT.type != bodhi or EVENT.id is not match("1.*"))
```

I would probably stick to string only, i.e. no lists, but all forms would work.

Reply:
@happz I have pushed an update, using a string for "when" conditions. However, I am facing some pre-commit issues I am not able to deal with.
{{{
newa/__init__.py:459: error: Unsupported left operand type for + ("None")  [operator]
newa/__init__.py:459: note: Left operand is of type "Optional[str]"
}}}
Really not sure how to properly approach this one.
{{{
newa/__init__.py:503: error: Extra argument "when" from **args for "Request"  [misc]
}}}
I though having "when" for a Request class is not necessary, however linter thinks we are passing this atribute as well.

Reply:
```
newa/__init__.py:459: error: Unsupported left operand type for + ("None") [operator]
newa/__init__.py:459: note: Left operand is of type "Optional[str]"

            elif key == 'when' and src["when"]:
                dest['when'] += f' and ( {src["when"]} )'
```

The warning is about the left side of the assignment, `dest['when']`. We know that `key` is `when`, and that `src['when']` is set and is true-ish, but `dest['when']` is still `Optional[str]`. Thanks to the previous branch we know it's set, but that's all, you still need to deal with `dest['when'] == None` or `dest['when'] == ''`. And `None + str` is what the complaint is about. Adding `dest['when']` check & handling for these would resolve the issue.

```
newa/init.py:503: error: Extra argument "when" from **args for "Request" [misc]
```

Correct. `**combination` - `combination` is of type `RawRecipeConfigDimension` and as such may contain the `when` key. You either need to teach `Request` to accept it, or, if that's not the right thing to do, drop it from `combination` before passing it to `Request`.

Reply:
Dropping it is what I am trying to do with `del combination["when"]` but it doesn't seem to work.

Reply:
Would it get better if you moved it one level to the left? I suspect it's hard for a linter to prove `when` won't be there, e.g. there's `if not test_result:` which may lead to `when` being still present.